pkg/migration/migration_test.go

```diff
@@ -23,8 +17,211 @@ import (
+
+	"github.com/libopenstorage/operator/drivers/storage/portworx/component"
+	"github.com/libopenstorage/operator/drivers/storage/portworx/manifest"
+	pxutil "github.com/libopenstorage/operator/drivers/storage/portworx/util"
+	corev1 "github.com/libopenstorage/operator/pkg/apis/core/v1"
+	"github.com/libopenstorage/operator/pkg/constants"
+	"github.com/libopenstorage/operator/pkg/controller/storagecluster"
+	testutil "github.com/libopenstorage/operator/pkg/util/test"
 )
 
+type ImageConfig struct {
+	CustomImageRegistry string
+	PortworxImage       string
+	Components          manifest.Release
+}
+
+func TestImageMigration(t *testing.T) {
+	// Test without custom image registry, component images match spec gen.
+	// StorageCluster should not configure any component image.
+	dsImages := ImageConfig{
+		PortworxImage: "portworx/oci-monitor:2.9.1",
+		Components: manifest.Release{
+			Stork:     "openstorage/stork:2.7.0",
+			Autopilot: "portworx/autopilot:1.3.1",
+			Telemetry: "purestorage/ccm-service:3.0.9",
+		},
+	}
+
+	expectedStcImages := ImageConfig{
+		PortworxImage:       "portworx/oci-monitor:2.9.1",
+		CustomImageRegistry: "",
+		Components: manifest.Release{
+			Stork:     "",
+			Autopilot: "",
+			Telemetry: "",
+		},
+	}
+
+	testImageMigration(t, dsImages, expectedStcImages)
+
+	// Test without custom image registry, and component images do not match default images.
+	// StorageCluster should configure component images.
+	dsImages = ImageConfig{
+		PortworxImage: "portworx/oci-monitor:2.9.1",
+		Components: manifest.Release{
+			Stork:     "a/stork:2.7.0",
+			Autopilot: "b/autopilot:1.3.1",
+			Telemetry: "c/ccm-service:3.0.9",
+		},
+	}
+
+	expectedStcImages = ImageConfig{
+		PortworxImage:       "portworx/oci-monitor:2.9.1",
+		CustomImageRegistry: "",
+		Components: manifest.Release{
+			Stork:     "a/stork:2.7.0",
+			Autopilot: "b/autopilot:1.3.1",
+			Telemetry: "c/ccm-service:3.0.9",
+		},
+	}
+
+	testImageMigration(t, dsImages, expectedStcImages)
+
+	// Test with custom image registry. Component images match spec gen.
+	// StorageCluster should not configure any component image.
+	dsImages = ImageConfig{
+		PortworxImage: "test/oci-monitor:2.9.1",
+		Components: manifest.Release{
+			Stork:     "test/openstorage/stork:2.7.0",
+			Autopilot: "test/portworx/autopilot:1.3.1",
+			Telemetry: "test/purestorage/ccm-service:3.0.9",
+		},
+	}
+
+	expectedStcImages = ImageConfig{
+		PortworxImage:       "oci-monitor:2.9.1",
+		CustomImageRegistry: "test",
+		Components: manifest.Release{
+			Stork:     "",
+			Autopilot: "",
+			Telemetry: "",
+		},
+	}
+
+	testImageMigration(t, dsImages, expectedStcImages)
+
+	// Test with custom image registry. Component images do not match spec gen.
+	// StorageCluster should configure component image.
+	dsImages = ImageConfig{
+		PortworxImage: "test/oci-monitor:2.9.1",
+		Components: manifest.Release{
+			Stork:     "test/a/stork:2.7.0",
+			Autopilot: "test/b/autopilot:1.3.1",
```

Comment:
Can we also add a test where the versions don't match but the path is fine? This would be most common scenario with customer's spec.

Reply:
This test is still missing.

Reply:
because it is set in desiredImage entirely now, we no longer check version.